bench/Main.hs

```diff
@@ -0,0 +1,285 @@
+output = if verbose ?config then liftIO . putStrLn else (\_ -> pure ())
+
+---------------------------------------------------------------------------------------
+
+type Experiment = TextDocumentIdentifier -> Session Bool
+
+data Bench = forall setup.
+  Bench
+  { name :: !String,
+    samples :: !Natural,
+    benchSetup :: TextDocumentIdentifier -> Session setup,
+    experiment :: setup -> Experiment
+  }
+
+bench :: String -> Natural -> Experiment -> Bench
+bench name samples userExperiment = Bench {..}
+  where
+    experiment () = userExperiment
+    benchSetup _ = return ()
+
+benchWithSetup ::
+  String ->
+  Natural ->
+  (TextDocumentIdentifier -> Session p) ->
+  (p -> Experiment) ->
+  Bench
+benchWithSetup = Bench
+
+runBenchmarks :: HasConfig => [Bench] -> IO ()
+runBenchmarks benchmarks = do
+  results <- forM benchmarks $ \b -> (b,) <$> runBench b
+
+  forM_ results $ \(Bench {name, samples}, duration) ->
+    output $
+      "TOTAL "
+        <> name
+        <> " = "
+        <> showDuration duration
+        <> " ("
+        <> show samples
+        <> " repetitions)"
+
+  when (outputCSV ?config) $ do
+    putStrLn $ intercalate ", " $ map name benchmarks
+    putStrLn $ intercalate ", " $ map (showDuration . snd) results
+
+runBench :: HasConfig => Bench -> IO Seconds
+runBench Bench {..} = handleAny (\e -> print e >> return (-1))
+  $ runSessionWithConfig conf cmd lspTestCaps dir
+  $ do
+    doc <- openDoc exampleModulePath "haskell"
+    void (skipManyTill anyMessage message :: Session WorkDoneProgressEndNotification)
+
+    liftIO $ output $ "Running " <> name <> " benchmark"
+    userState <- benchSetup doc
+    let loop 0 = return True
+        loop n = do
+          (t, res) <- duration $ experiment userState doc
+          if not res
+            then return False
+            else do
+              output (showDuration t)
+              loop (n -1)
+
+    (t, res) <- duration $ loop samples
+
+    exitServer
+    -- sleeep to give ghcide a chance to print the RTS stats
+    liftIO $ threadDelay 50000
+
+    return $ if res then t else -1
+  where
+    cmd =
+      unwords $
+        [ "ghcide",
+          "--lsp",
+          "--cwd",
+          dir,
+          "+RTS",
+          "-N4",
+          "-S",
+          "-RTS"
+        ]
+          ++ concat
+            [ ["--shake-profiling", path]
+              | Just path <- [shakeProfiling ?config]
+            ]
+    dir = "bench/example/" <> examplePackage
+    lspTestCaps =
+      fullCaps {_window = Just $ WindowClientCapabilities $ Just True}
+    conf =
+      defaultConfig
+        { logStdErr = verbose ?config,
+          logMessages = False,
+          logColor = False
+        }
+
+setup :: HasConfig => IO ()
+setup = do
+  alreadyExists <- doesDirectoryExist "bench/example"
+  when alreadyExists $ removeDirectoryRecursive "bench/example"
+  callCommand $ "cabal unpack -v0 " <> examplePackage <> " -d bench/example"
```

Comment:
`cabal get` replaced `cabal unpack` I think. https://github.com/haskell/cabal/issues/5410

Reply:
Thanks, I wasn't aware of that